### PR TITLE
Readme

### PR DESCRIPTION
**Pull request by ryanwholey**
Adds the plugin basics in the readme. This could be a good time to discuss any changes we want to see in the templating portion of the annotations.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,9 +1,170 @@
...
+| `exec-forward.pod.kubernetes.io/post-connect` | A JSON formatted list of commands executed after establishing a port-forwarding connection |
+| `exec-forward.pod.kubernetes.io/command` | A single JSON formatted command ran after `post-connect` |
+
+#### Command
+
+##### Object
+
+The command object represents an exec command and associated configuration to be executed by the plugin during the port-forwarding lifecycle.
+
+| Attribute | Description | Optional |
```
Comment (bendrucker) [inline README.md:55]: Invert: Optional -> Required

File: README.md [hunk 2]
```diff
@@ -1,9 +1,170 @@
...
+| --- | --- | --- |
+| `id` | A unique identifier that can be used in subsequent commands to reference a previous command's output | `true` |
+| `command` | The command to run as an array of strings | `false` |
+| `interactive` | Whether the command should be run in interactive mode. Default is `false`. Note: the main `command` is always run in interactive mode | `true` |
+| `name` | The display name for the command, shown during execution | `false` |
+
+##### Rendering
+
+Exec commands are [Go templated](https://pkg.go.dev/text/template) using inputs from the CLI, the `args` annotation, configuration from the port-forwarding connection and previous command outputs.
```
Comment (bendrucker) [inline README.md:64]:
```suggestion
Exec commands are rendered via [Go templates](https://pkg.go.dev/text/template) using inputs from the CLI, the `args` annotation, configuration from the port-forwarding connection and previous command outputs.
```

File: README.md [hunk 3]
```diff
@@ -1,9 +1,170 @@
...
+
+| Namespace | Description | Example |
+|---|---|---|
+| `.Args` | Arguments read from the `args` annotation and overridden using the `--arg\|-a` CLI flags | `{{.Args.username}}` |
+| `.Outputs` | Stdout and Stderr from previously ran commands, stored by command `id` | `{{.Outputs.foo.Stdout}}` |
```
Comment (bendrucker) [inline README.md:69]: Outputs could probably just be stdout straight up, i.e. `.Outputs.foo`. Stderr can be stored to print errors but doesn't need to be available for command templating.

File: README.md [hunk 4]
```diff
@@ -1,9 +1,170 @@
...
+      spec:
+        containers:
+          ...
+          - ports:
+              name: http
+              ...
+```
+
+```sh
+kubectl port-forward deployment/foo http
```
Comment (bendrucker) [inline README.md:169]: Wrong command

File: README.md [hunk 5]
```diff
@@ -1,9 +1,170 @@
...
+        "command":[
+          "psql",
+          "postgres://{{.Args.username}}:{{ trim .Outputs.password.Stdout | urlquery | sensitive }}@localhost:{{.Config.LocalPort}}/db"
+        ]
+      }
+spec:
+  containers:
+    image: alpine/socat
+    - command:
+      - socat
```
Comment (bendrucker) [inline README.md:118]: Worth explaining the use of `socat` a bit more here and how it's serving as an active forwarder in Kubernetes.

File: README.md [hunk 6]
```diff
@@ -1,9 +1,170 @@
...
+
+##### Rendering
+
+Exec commands are [Go templated](https://pkg.go.dev/text/template) using inputs from the CLI, the `args` annotation, configuration from the port-forwarding connection and previous command outputs.
+
+| Namespace | Description | Example |
+|---|---|---|
+| `.Args` | Arguments read from the `args` annotation and overridden using the `--arg\|-a` CLI flags | `{{.Args.username}}` |
+| `.Outputs` | Stdout and Stderr from previously ran commands, stored by command `id` | `{{.Outputs.foo.Stdout}}` |
+| `.Config` | Port-forwarding configuration | `{{.Config.LocalPort }}` |
```
Comment (bendrucker) [inline README.md:70]: `.Config` is a bit overloaded, maybe just `.LocalPort` or `.Ports.Local`. The normal case should be to choose a local port at random, in which case the value isn't user-provided config.

File: README.md [hunk 7]
```diff
@@ -1,9 +1,170 @@
...
+
+##### Object
+
+The command object represents an exec command and associated configuration to be executed by the plugin during the port-forwarding lifecycle.
+
+| Attribute | Description | Optional |
+| --- | --- | --- |
+| `id` | A unique identifier that can be used in subsequent commands to reference a previous command's output | `true` |
+| `command` | The command to run as an array of strings | `false` |
+| `interactive` | Whether the command should be run in interactive mode. Default is `false`. Note: the main `command` is always run in interactive mode | `true` |
```
Comment (bendrucker) [inline README.md:59]:
```suggestion
| `interactive` | Whether the command should be run in interactive mode and can receive user input. Default is `false`. Note: the main `command` is always run in interactive mode | `true` |
```

File: README.md (hunk 1 above)
Comment (bendrucker) [reply README.md:55]: Add a default column

File: README.md [hunk 8]
```diff
@@ -1,9 +1,170 @@
...
+| Flag | Short | Description | Default |
+|---|---|---|---|
+| `--arg` | `-a` | `key=value` arguments passed to commands | `[]` |
+| `--verbose` |`-v`| Whether to log verbosely |`false` |
+| `--pod-timeout` | `-t` | Time to wait for an attachable pod to become available | `500` (ms) |
+| `--persist` | `-p` | Whether to persist the forwarding connection after the main command has finished | `false` |
+
+## Administration
+
+Administrators can store complex behavior in Kubernetes pod annotations, allowing users to run a single `kubectl` to interact with remote resources.
```
Comment (bendrucker) [inline README.md:30]:
```suggestion
Administrators can store complex behavior in Kubernetes pod annotations, allowing users to run a single `kubectl` command to interact with remote resources.
```

File: README.md (hunk 3 above)
Comment (ryanwholey) [reply README.md:69]: yeah i was thinking about this. I'll open a separate PR.

File: README.md (hunk 6 above)
Comment (ryanwholey) [reply README.md:70]: Yeah agreed, the other info on Config is not useful, I'll open a separate PR. The other similar option I've had my eye on is address (default localhost), which could also be passed at top level

**Review by bendrucker — APPROVED** [review]

**Pull request merged**